src/main/java/com/microsoft/graph/httpcore/RetryHandler.java

```diff
@@ -1,110 +1,142 @@
     
-    private final int MSClientErrorCodeTooManyRequests = 429;
-    private final int MSClientErrorCodeServiceUnavailable  = 503;
-    private final int MSClientErrorCodeGatewayTimeout = 504;
-    private final RetryOptions mRetryOption;
+    public static final int MSClientErrorCodeTooManyRequests = 429;
+    public static final int MSClientErrorCodeServiceUnavailable  = 503;
+    public static final int MSClientErrorCodeGatewayTimeout = 504;
     
-    public RetryHandler(RetryOptions option) {
-        super();
-        this.mRetryOption = option;
+    private final long DELAY_MILLISECONDS = 1000;
+    
+    /*
+     * @retryOption Create Retry handler using retry option
+     */
+    public RetryHandler(RetryOptions retryOption) {
+        this.mRetryOption = retryOption;
+        if(this.mRetryOption == null) {
+        	this.mRetryOption = new RetryOptions();
+        }
     }
-
+    /*
+     * Initialize retry handler with default retry option
+     */
     public RetryHandler() {
         this(null);
     }
     
-	@Override
-	public boolean retryRequest(HttpResponse response, int executionCount, HttpContext context) {
+	boolean retryRequest(Response response, int executionCount, Request request, RetryOptions retryOptions) {
 		
-		RetryOptions retryOption = (RetryOptions)context.getAttribute(MiddlewareType.RETRY.toString());
-		if(retryOption != null) {
-			return retryOption.shouldRetry().shouldRetry(response, executionCount, context);
+		// Should retry option
+		// Use should retry common for all requests
+		IShouldRetry shouldRetryCallback = null;
+		if(retryOptions != null) {
+			shouldRetryCallback = retryOptions.shouldRetry();
 		}
-		if(mRetryOption != null) {
-			return mRetryOption.shouldRetry().shouldRetry(response, executionCount, context);
+		// Call should retry callback
+		if(shouldRetryCallback != null && !shouldRetryCallback.shouldRetry(response, executionCount, request, retryOptions.delay())) {
+				return false;
 		}
 		
 		boolean shouldRetry = false;
-		int statusCode = response.getStatusLine().getStatusCode();
-		shouldRetry = (executionCount < maxRetries) && checkStatus(statusCode) && isBuffered(response, context);
+		// Status codes 429 503 504
+		int statusCode = response.code();
+		// Only requests with payloads that are buffered/rewindable are supported. 
+		// Payloads with forward only streams will be have the responses returned 
+		// without any retry attempt.
+		shouldRetry = (executionCount <= retryOptions.maxRetries()) && checkStatus(statusCode) && isBuffered(response, request);
 		
 		if(shouldRetry) {
-			Header header = response.getFirstHeader(RETRY_AFTER);
-			if(header != null) 
-				retryInterval = Long.parseLong(header.getValue());
-			else
-				retryInterval = (long)Math.pow(2.0, (double)executionCount) * DELAY_MILLISECONDS;
+			long retryInterval = getRetryAfter(response, retryOptions.delay(), executionCount);
+			try {
+				Thread.sleep(retryInterval);
```

Comment:
check how apache httpclient has implemented retry after mechanism.

Reply:
apache also uses same Thread.sleep mechanism for waiting between retries -[ServiceUnavailableRetryExec](https://github.com/apache/httpcomponents-client/blob/7f5668797a964a63fed8f03e1ea660dd8cd9e232/httpclient5/src/main/java/org/apache/hc/client5/http/impl/classic/ServiceUnavailableRetryExec.java#L99)